engine/runtime-compiler/src/main/java/org/enso/compiler/pass/MiniPassTraverser.java

```diff
@@ -0,0 +1,61 @@
+package org.enso.compiler.pass;
+
+import org.enso.compiler.context.ModuleContext;
+import org.enso.compiler.core.CompilerError;
+import org.enso.compiler.core.IR;
+import org.enso.compiler.core.ir.Expression;
+import org.enso.compiler.core.ir.Module;
+
+public final class MiniPassTraverser {
+  private MiniPassTraverser() {}
+
+  /**
+   * Compiles the whole module with chained {@link MiniIRPass}.
+   *
+   * @param miniPass Mini pass that must support module compilation, i.e., should be created with
+   *     {@link MiniPassFactory#createForModuleCompilation(ModuleContext)}.
+   * @return Transformed module IR.
+   */
+  public static Module compileModuleWithMiniPass(Module moduleIr, MiniIRPass miniPass) {
+    var newModuleIr = compileRecursively(moduleIr, miniPass);
+    return (Module) newModuleIr;
+  }
+
+  public static Expression compileInlineWithMiniPass(Expression exprIr, MiniIRPass miniPass) {
+    var preparedMiniPass = miniPass.prepare(exprIr);
+    var newIr = compileRecursively(exprIr, preparedMiniPass);
+    if (!preparedMiniPass.checkPostCondition(newIr)) {
+      throw new CompilerError("Post condition failed after applying mini pass " + preparedMiniPass);
+    }
+    return (Expression) newIr;
+  }
+
+  private static IR compileRecursively(IR ir, MiniIRPass miniPass) {
+    var preparedMiniPass = miniPass.prepare(ir);
+    IR newIr;
+    if (ir.children().isEmpty()) {
+      newIr = ir;
+    } else {
+      var transformedChildren =
+          ir.children()
+              .map(
```

Comment:
I was looking at [ForkJoinPool](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/ForkJoinTask.html). They claim the best way to _split the work_ is

>  As a very rough rule of thumb, a task should perform more than 100 and less than 10000 basic computational steps, and should avoid indefinite looping"

Maybe we should really be traversing only `ir.Expression`. Maybe the other elements like `CallArgument` are too small to be processed by own tasks.

```suggestion
              filter(_.isInstanceOf[Expression]).map(
```

Reply:
I really believe we can achieve our performance/memory goals with `IR.mapExpression`. Here is a tag [jtulach/MiniPassTraverserWithMapChildren](https://github.com/JaroslavTulach/enso/releases/tag/jtulach/MiniPassTraverserWithMapChildren) that demonstrates it. I'll put it into this branch as well - but feel free to remove it, if we decide we desperately need `withNewChildren` for some pass.

Tests seem to work. Prior to my commit there were some failures in `org.enso.compiler.test.pass.analyse.TailCallTest` - after my commit there are also only failures in `TailCallTest`. Thus I assume all the other _mini passes_ are working fine with the `mapExpressions` approach.

Reply:
I am struggling with implementing `TailCall` mini pass with this pattern. `TailCall` is a pass that traverses *the whole IR tree* and attaches to every node either `TAIL` or `NOT_TAIL`.

Following is a picture of IR tree with numbered `prepare` and `transform` steps of `MiniIRPassTraverser` (implemented with `mapExpressions` as contributed by @JaroslavTulach on revision https://github.com/enso-org/enso/pull/11191/commits/ce56abb7e0e449bded3cf74e63a617ff448c9898):

![tmp dot](https://github.com/user-attachments/assets/9206dfd3-b05e-4aed-9bb6-55fd7c6a384c)


As you can see, the `MiniIRPassTraverser` basically ignores the whole `bindings[1]` and `bindings[2]` subtrees, because they are not Expressions. How am I supposed to transform those? Should I manually recursively traverse all the children of Expression and look for IR elements that are not Expressions and do my transformation of those? If yes, then this is would not be a *mini pass* at all, it would be a weird monstrosity somewhere between a mega pass and mini pass. I am not even mentioning how ugly (difficult to maintain) and non-intuitive it is to work with this.

In my opinion, **this is a proof that the solution with traversing only Expressions with `mapExpressions` is not sustainable**, and we should not continue in this direction.

Reply:
Last Friday we _agreed to disagree_. This PR is going to work with `mapExpressions` and the fact it is green is a proof it can work that way.

Reply:
On the other hand, while implementing the `TailCall` pass, I came to an agreement that the `prepare` method shall get _an edge_ in the `IR` tree - e.g. its signature should be `MiniIRPass prepare(IR parent, IR child)`. _Traversing by edges_ is however **orthogonal** to using `mapExpressions` or `withNewChildren` question.

Reply:
Isn't `TailCall` pass now simplified (https://github.com/enso-org/enso/commit/d9b3c75c5d9440db85579dbb8e8f35a3f876cda8) and therefore not a representative case for this problem? Meaning we will likely encounter it in other passes?

Reply:
> Isn't TailCall pass now simplified (https://github.com/enso-org/enso/commit/d9b3c75c5d9440db85579dbb8e8f35a3f876cda8)

No, that change has no effect on the traversal. `TailCall` is as representative as it was and is the main reason why I see that [edge traversal](https://github.com/enso-org/enso/pull/11191/commits/ffd27dfe9b6fbbd59cccb35633cb802be097815c) as advocated by Pavel is more meaningful than _vertex traversal_.

> likely encounter it in other passes?

It remains to be seen what we will encounter in other passes. However I'd like to [highlight following quote](https://github.com/enso-org/enso/pull/11191#discussion_r1787612365): _"by using `mapExpressions` ... we get better mapping of current `IRPass` implementations"_ that are already using `mapExpressions` to traverse!